src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,426 @@
+                    xmlList.Add(xmlPath);
+                }
+                else if (compare > 0) {
+                    pdbCnt++;
+                    peCnt--;
+                }
+                if (pdbCnt == pdbFiles.Count) break;
+            }
+            return xmlList;
+        }
+
+        private static List<string> GrabFiles(List<string> paths, string searchPattern)
+        {
+            List<string> files = new List<string>();
+            foreach (var assemDir in paths)
+            {
+                if (Directory.Exists(assemDir))
+                {
+                    foreach (var peFile in Directory.GetFiles(assemDir, searchPattern, SearchOption.AllDirectories))
+                    {
+                        if (!peFile.Contains(".ni.dll"))
+                        {
+                            files.Add(peFile);
+                        }
+                    }
+                }
+            }
+            return files;
+        }
+
+        private static void GenXmlFromPdb(string assemblyPath, string pdbPath, string xmlPath)
+        {
+            using var peStream = new FileStream(assemblyPath, FileMode.Open, FileAccess.Read);
+            using var pdbStream = new FileStream(pdbPath, FileMode.Open, FileAccess.Read);
+            using var dstFileStream = new FileStream(xmlPath, FileMode.Create, FileAccess.ReadWrite);
+            using var sw = new StreamWriter(dstFileStream, Encoding.UTF8);
+            PdbToXmlOptions options = PdbToXmlOptions.ResolveTokens | PdbToXmlOptions.IncludeTokens;
+
+            PdbToXmlConverter.ToXml(sw, pdbStream, peStream, options);
+        }
+
+        private static void RemoveTempDirectory()
+        {
+            if (Directory.Exists(tempDirectory))
+            {
+                Directory.Delete(tempDirectory, true);
+            }
+        }
+
+        private static void Symbolicator(IConsole console, List<string> xmlList, string inputPath, string outputPath)
+        {
+            if (xmlList.Count == 0)
+            {
+                RemoveTempDirectory();
+                throw new FileNotFoundException("Xml file not found\n");
+            }
+
+            GetLineFromStack(console, xmlList, inputPath, outputPath);
+
+            RemoveTempDirectory();
+        }
+
+        internal sealed class StackTraceInfo
+        {
+            public string Type;
+            public string Method;
+            public string Param;
+            public string Assembly;
+            public string Token;
+            public string Offset;
+            public string Document;
+            public string Filepath;
+            public string Filename;
+            public string StartLine;
+            public string EndLine;
+        }
+
+        private static void GetLineFromStack(IConsole console, List<string> xmlList, string inputPath, string outputPath)
+        {
+            try
+            {
+                StreamWriter fswi = null;
```

Comment:
nit: could we name these something more verbose? It's difficult to grok what these `Stream{Reader|Writer}`s are pointed at from their names and they're all very similar to each other.